### ISSUE TITLE
Unpacking C-style arrays size Issue

### ISSUE THREAD

**Issue by mowaiskhan** [msgpack/msgpack-c#485]
Hello,

I observed something while working with C-Style arrays for char.

It crashes while unpacking if you have used the whole array like below.

char array[100];

for(int i = 0;i < 100;i++)
   array[i] = '0';

But, it will work fine if you declare above array a bit bigger like char array[150] and don't write bytes beyond 100 bytes for example.

Check below code for a working sample.

http://melpon.org/wandbox/permlink/I3kLQTPNPJlsJh3T

Thanks & Regards,
Owais.

**Comment by redboltz** [comment]
msgpack-c considers char[] as a null terminated string by default. You need to set the last char to null.

If you want to char[] as array, use array_ref.

**Comment by redboltz** [comment]
My previous comment might be incorrect. I will check it again later. I'm on a business trip now. I will back the day after tomorrow.

**Comment by mowaiskhan** [comment]
Sure, Thanks. I am using it on a branch you created for my previous request #465

**Comment by redboltz** [comment]
> Sure, Thanks. I am using it on a branch you created for my previous request #465

Thank you for the information. BTW, #466 that fixed #465 has been merged to the master.

You can see the type mapping In https://github.com/msgpack/msgpack-c/pull/466 comments:

```
Existing mapping:
convert:
pack:
char[N] => STR
const char[N] => STR
object:
char[N] => STR (v1 only)
const char[N] => STR (v1 only)
object_with_zone:
char[N] => STR
const char[N] => STR

Additional mapping:

convert:
STR => char[N]
ARRAY => T[N]
pack:
T[N] => ARRAY
const T[N] => ARRAY
object:
object_with_zone:
T[N] => ARRAY
const T[N] => ARRAY
```

I updated your code, then error is disappeard with array size 100.
See line 29

http://melpon.org/wandbox/permlink/euQ1jw42SdDFjZxU

Let me describe the design decision. It is not reflected the wiki document yet.

Consider the following data.

```C++
char data[] = "ABC"; 
```

The size of data is 4. It contains, A, B, C, and '\0'. It is packed as msgpack format STR, and its length is 3, contains A, B, C. I believe that it is an expected behavior for many users. If STR contains '\0', it is difficult to inter-operate other types and other languages.

STR means https://github.com/msgpack/msgpack/blob/master/spec.md#formats-str

When packing, msgpack-c expects the C-style array of char is '\0' terminated. When unpacking, msgpack-c add '\0' after the last character. So the size of array should have STR size + 1 at least.

Consider the following data:
```C++
char data[] = {'A', 'B', 'C' }; 
```
The size of data is 3. It contains, A, B, and C. It isn't '\0' terminated, so that cause error.

I personally don't recommend using C-Style array. C++ provides better alternatives. You can use std::string for strings, and std::array for arrays. You can also use `std::array<char, N>' for string. In this case, you don't need to add `\0` to the tail.

See http://melpon.org/wandbox/permlink/aNbCbuqSI6iRNCKM

Legacy C-Style array has many restrictions. I need to care C style behavior as follows:

```C++
char data[] = "ABC"; // pack to STR "ABC", and convert to "ABC\0".
```

**Comment by redboltz** [comment]
I come up with a better approach.

In the case the target type is char[] or const char[],

If the array is '\0' teminated, msgpack-c packs the characters before
'\0', otherwise packs all characters.

When converting, the array has the size that is greater than
msgpack::object STR's size, msgpack-c adds '\0' to just after converted
characters. Otherwise msgpack-c doesn't add '\0'.

I will implement this.

**Comment by redboltz** [comment]
@mowaiskhan, I sent the PR #488 
Could you test it on your environment?

Here is the instruction:
```
git checkout -b redboltz-fix_485 master
git pull https://github.com/redboltz/msgpack-c.git fix_485
```

**Comment by mowaiskhan** [comment]
Sorry for a late reply, I just read the thread and verified PR #488 

I tested it for

char singlechar;
char nonnullarray[30];
char nullarray[30];

It works fine for all the cases. Thanks a lot for change as I am incorporating msgpack-c in legacy code where I have non-null char arrays.

Saw the code as well, it now starts making some sense to me :)

**Comment by redboltz** [comment]
I merged #488. 
@mowaiskhan , I found the better way to treat legacy array of chars thanks to your cooperation.

**Comment by mowaiskhan** [comment]
You are welcome, Thanks.
